File: union.lua
local M = {}

local function rect_order(bl1, tr1, bl2, tr2)
    if bl1 > tr2 or bl2 > tr2 then return "NOINTERSECTION" end
    if bl1  < bl2 and tr1  > tr2 then return "OUTER" end
    if bl2  < bl1 and tr2  > tr1 then return "INNER" end
    if bl1 == bl2 and tr1 == tr2 then return "EQUAL" end
    if bl1 == bl2 and tr1  < tr2 then return "HALFEQUALLEFTREGULAR" end
    if bl1 == bl2 and tr1  > tr2 then return "HALFEQUALLEFTINVERSE" end
    if tr1 == tr2 and bl1  < bl2 then return "HALFEQUALRIGHTREGULAR" end
    if tr1 == tr2 and bl1  > bl2 then return "HALFEQUALRIGHTINVERSE" end
    if tr1 >= bl2 and bl1  < bl2 then return "REGULAR" end
    if tr2 >= bl1 and bl2  < bl1 then return "INVERSE" end
    return "NOINTERSECTION"
end

local function rectangle_union(rect1, rect2)
    local bl1x, bl1y = rect1.bl:unwrap()
    local tr1x, tr1y = rect1.tr:unwrap()
    local bl2x, bl2y = rect2.bl:unwrap()
    local tr2x, tr2y = rect2.tr:unwrap()
    local xorder = rect_order(bl1x, tr1x, bl2x, tr2x)
    if xorder == "NOINTERSECTION" then
        return nil
    end
    local yorder = rect_order(bl1y, tr1y, bl2y, tr2y)
    if yorder == "NOINTERSECTION" then
        return nil
    end
    if xorder == "OUTER" and (yorder == "HALFEQUALLEFTINVERSE" or yorder == "HALFEQUALRIGHTREGULAR") then
        return { bl = rect1.bl:copy(), tr = rect1.tr:copy() }
    end
    if xorder ~= "EQUAL" and yorder ~= "EQUAL" then -- polygon union, one order has to be EQUAL for rectangle union
        return nil
    end
    local blx, bly, trx, try = 0, 0, 0, 0
    if xorder == "HALFEQUALLEFTREGULAR" then
        blx = bl1x
        trx = tr2x
    elseif xorder == "HALFEQUALLEFTINVERSE" then
        blx = bl1x
        trx = tr1x
    elseif xorder == "HALFEQUALRIGHTREGULAR" then
        blx = bl1x
        trx = tr1x
    elseif xorder == "HALFEQUALRIGHTINVERSE" then
        blx = bl2x
        trx = tr1x
    elseif xorder == "EQUAL" then
        blx = bl1x
        trx = tr1x
    elseif xorder == "OUTER" then
        blx = bl1x
        trx = tr1x
    elseif xorder == "INNER" then
        blx = bl2x
        trx = tr2x
    elseif xorder == "REGULAR" then
        blx = bl1x
        trx = tr2x
    else
        blx = bl2x
        trx = tr1x
    end
    if yorder == "HALFEQUALLEFTREGULAR" then
        bly = bl1y
        try = tr2y
    elseif yorder == "HALFEQUALLEFTINVERSE" then
        bly = bl1y
        try = tr1y
    elseif yorder == "HALFEQUALRIGHTREGULAR" then
        bly = bl1y
        try = tr1y
    elseif yorder ==  "HALFEQUALRIGHTINVERSE" then
        bly = bl2y
        try = tr1y
    elseif yorder == "EQUAL" then
        bly = bl1y
        try = tr1y
    elseif yorder == "OUTER" then
        bly = bl1y
        try = tr1y
    elseif yorder == "INNER" then
        bly = bl2y
        try = tr2y
    elseif yorder == "REGULAR" then
        bly = bl1y
        try = tr2y
    else
        bly = bl2y
        try = tr1y
    end
    return { bl = point.create(blx, bly), tr = point.create(trx, try) }
end

function M.rectangle_all(rectangles)
    local i = 1
    local j = 2
    while #rectangles > 1 do
        if i == #rectangles and j == #rectangles + 1 then break end
        local rect1 = rectangles[i]
        local rect2 = rectangles[j]
        local result = rectangle_union(rect1, rect2)
        if result then
            rectangles[i] = result
            table.remove(rectangles, j)
            -- restart iteration
            i = 1
            j = 2
        else
            j = j + 1
            if j > #rectangles then
                i = i + 1
                j = i + 1
            end
        end
    end
end

return M

File: src/union.c
#include "union.h"

#include <stddef.h>
#include <stdbool.h>

#include "point.h"
#include "shape.h"

enum order {
    NOINTERSECTION,
    REGULAR,
    INVERSE,
    HALFEQUALLEFTREGULAR,
    HALFEQUALLEFTINVERSE,
    HALFEQUALRIGHTREGULAR,
    HALFEQUALRIGHTINVERSE,
    EQUAL,
    OUTER,
    INNER
};

static enum order rect_order(coordinate_t bl1, coordinate_t tr1, coordinate_t bl2, coordinate_t tr2)
{
    if(bl1  > tr2 || bl2  > tr1) return NOINTERSECTION;
    if(bl1  < bl2 && tr1  > tr2) return OUTER;
    if(bl2  < bl1 && tr2  > tr1) return INNER;
    if(bl1 == bl2 && tr1 == tr2) return EQUAL;
    if(bl1 == bl2 && tr1  < tr2) return HALFEQUALLEFTREGULAR;
    if(bl1 == bl2 && tr1  > tr2) return HALFEQUALLEFTINVERSE;
    if(tr1 == tr2 && bl1  < bl2) return HALFEQUALRIGHTREGULAR;
    if(tr1 == tr2 && bl1  > bl2) return HALFEQUALRIGHTINVERSE;
    if(tr1 >= bl2 && bl1  < bl2) return REGULAR;
    if(tr2 >= bl1 && bl2  < bl1) return INVERSE;
    return NOINTERSECTION;
}

struct shape* rectangle_shape_union(struct shape* rect1, struct shape* rect2)
{
    const struct point *bl1;
    const struct point *tr1;
    shape_get_rectangle_points(rect1, &bl1, &tr1);
    const struct point *bl2;
    const struct point *tr2;
    shape_get_rectangle_points(rect2, &bl2, &tr2);
    enum order xorder = rect_order(bl1->x, tr1->x, bl2->x, tr2->x);
    enum order yorder = rect_order(bl1->y, tr1->y, bl2->y, tr2->y);
    if(xorder == NOINTERSECTION || yorder == NOINTERSECTION)
    {
        return NULL;
    }
    if(xorder != EQUAL && yorder != EQUAL) // polygon union, one order has to be EQUAL for rectangle union
    {
        return NULL;
    }
    coordinate_t blx, bly, trx, try;
    switch(xorder)
    {
        case HALFEQUALLEFTREGULAR:
            blx = bl1->x;
            trx = tr2->x;
            break;
        case HALFEQUALLEFTINVERSE:
            blx = bl1->x;
            trx = tr1->x;
            break;
        case HALFEQUALRIGHTREGULAR:
            blx = bl1->x;
            trx = tr1->x;
            break;
        case HALFEQUALRIGHTINVERSE:
            blx = bl2->x;
            trx = tr1->x;
            break;
        case EQUAL:
            blx = bl1->x;
            trx = tr1->x;
            break;
        case OUTER:
            blx = bl1->x;
            trx = tr1->x;
            break;
        case INNER:
            blx = bl2->x;
            trx = tr2->x;
            break;
        case REGULAR:
            blx = bl1->x;
            trx = tr2->x;
            break;
        case INVERSE:
            blx = bl2->x;
            trx = tr1->x;
            break;
        default: // silence warning about not handling NOINTERSECTION, which is handled earlier
            return NULL;
            break;
    }
    switch(yorder)
    {
        case HALFEQUALLEFTREGULAR:
            bly = bl1->y;
            try = tr2->y;
            break;
        case HALFEQUALLEFTINVERSE:
            bly = bl1->y;
            try = tr1->y;
            break;
        case HALFEQUALRIGHTREGULAR:
            bly = bl1->y;
            try = tr1->y;
            break;
        case HALFEQUALRIGHTINVERSE:
            bly = bl2->y;
            try = tr1->y;
            break;
        case EQUAL:
            bly = bl1->y;
            try = tr1->y;
            break;
        case OUTER:
            bly = bl1->y;
            try = tr1->y;
            break;
        case INNER:
            bly = bl2->y;
            try = tr2->y;
            break;
        case REGULAR:
            bly = bl1->y;
            try = tr2->y;
            break;
        case INVERSE:
            bly = bl2->y;
            try = tr1->y;
            break;
        default: // silence warning about not handling NOINTERSECTION, which is handled earlier
            return NULL;
            break;
    }
    struct shape* new = shape_create_rectangle(shape_get_layer(rect1), blx, bly, trx, try);
    return new;
}

int rectangle_union(coordinate_t blx1, coordinate_t bly1, coordinate_t trx1, coordinate_t try1, coordinate_t blx2, coordinate_t bly2, coordinate_t trx2, coordinate_t try2, coordinate_t* blx, coordinate_t* bly, coordinate_t* trx, coordinate_t* try)
{
    enum order xorder = rect_order(blx1, trx1, blx2, trx2);
    enum order yorder = rect_order(bly1, try1, bly2, try2);
    if(xorder == NOINTERSECTION || yorder == NOINTERSECTION)
    {
        return 0;
    }
    if(xorder != EQUAL && yorder != EQUAL) // polygon union, one order has to be EQUAL for rectangle union
    {
        return 0;
    }
    switch(xorder)
    {
        case HALFEQUALLEFTREGULAR:
            *blx = blx1;
            *trx = trx2;
            break;
        case HALFEQUALLEFTINVERSE:
            *blx = blx1;
            *trx = trx1;
            break;
        case HALFEQUALRIGHTREGULAR:
            *blx = blx1;
            *trx = trx1;
            break;
        case HALFEQUALRIGHTINVERSE:
            *blx = blx2;
            *trx = trx1;
            break;
        case EQUAL:
            *blx = blx1;
            *trx = trx1;
            break;
        case OUTER:
            *blx = blx1;
            *trx = trx1;
            break;
        case INNER:
            *blx = blx2;
            *trx = trx2;
            break;
        case REGULAR:
            *blx = blx1;
            *trx = trx2;
            break;
        case INVERSE:
            *blx = blx2;
            *trx = trx1;
            break;
        default: // silence warning about not handling NOINTERSECTION, which is handled earlier
            return 0;
            break;
    }
    switch(yorder)
    {
        case HALFEQUALLEFTREGULAR:
            *bly = bly1;
            *try = try2;
            break;
        case HALFEQUALLEFTINVERSE:
            *bly = bly1;
            *try = try1;
            break;
        case HALFEQUALRIGHTREGULAR:
            *bly = bly1;
            *try = try1;
            break;
        case HALFEQUALRIGHTINVERSE:
            *bly = bly2;
            *try = try1;
            break;
        case EQUAL:
            *bly = bly1;
            *try = try1;
            break;
        case OUTER:
            *bly = bly1;
            *try = try1;
            break;
        case INNER:
            *bly = bly2;
            *try = try2;
            break;
        case REGULAR:
            *bly = bly1;
            *try = try2;
            break;
        case INVERSE:
            *bly = bly2;
            *try = try1;
            break;
        default: // silence warning about not handling NOINTERSECTION, which is handled earlier
            return 0;
            break;
    }
    return 1;
}

size_t union_rectangle_all(struct vector* rectangles)
{
    int i = 0;
    int j = 1;
    while(1)
    {
        //if(i == vector_size(rectangles) - 1 && j == vector_size(rectangles)) break;
        if(i >= (int)vector_size(rectangles) - 1) break;
        struct shape* rect1 = vector_get(rectangles, i);
        struct shape* rect2 = vector_get(rectangles, j);
        struct shape* result = rectangle_shape_union(rect1, rect2);
        if(result)
        {
            vector_set(rectangles, i, result);
            shape_destroy(rect1);
            vector_remove(rectangles, j);
            // restart iteration
            i = 0;
            j = 1;
        }
        else
        {
            ++j;
        }
        if(j > (int)vector_size(rectangles) - 1)
        {
            ++i;
            j = i + 1;
        }
    }
    return vector_size(rectangles);
}

